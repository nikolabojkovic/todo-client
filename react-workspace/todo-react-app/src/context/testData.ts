import { 
  ITodo,
  IPaging, 
  ISort, 
  SortDirection, 
  IFilter,
  StateFilter, 
  IGeneralSettings, 
  IPaginationSettings, 
  ISearchSettings, 
  ISettings, 
  ListContainerType, 
  PaginationType, 
  BackgroundColor,
  IThemeSettings,
  BsThemes
} from "../models";
import { DisplayMode, IState } from "./";

export const todoListTestData: ITodo[] = [
  {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2022, 1, 4)
  } as ITodo,
  {
    id: 2,
    title: "Task 2",
    description: "Description 2",
    completed: true,
    createdAt: new Date(2022, 1, 5)
  } as ITodo,
  {
    id: 3,
    title: "Task 3",
    description: "Description 3",
    completed: false,
    createdAt: new Date(2022, 1, 6)
  } as ITodo,
  {
    id: 4,
    title: "Task 4",
    description: "Description 4",
    completed: false,
    createdAt: new Date(2022, 1, 7)
  } as ITodo,
  {
    id: 6,
    title: "Task 6",
    description: "Description 6",
    completed: false,
    createdAt: new Date(2022, 2, 4)
  } as ITodo,
  {
    id: 5,
    title: "Task 5",
    description: "Description 5",
    completed: false,
    createdAt: new Date(2022, 3, 8)
  } as ITodo
] as ITodo[];

export const stateTestData: IState = 
{
  originalList: todoListTestData,
  displayList: todoListTestData,
  displayMode: DisplayMode.All,
  search: { searchTerm: '' },
  filter: { state: StateFilter.all } as IFilter,
  sort: { column: 'createdAt', direction: SortDirection.Asc} as ISort,
  paging: {
    totalCount: 6,
    activePage: 1,
    itemsPerPage: 5,
    startIndex: 0,
    endIndex: 5,
  } as IPaging,
  settings: {
    general: {
      isConfirmEnabled: true,
      isPaginationEnabled: true,
      isInfiniteScrollEnabled: false,
      listSizeType: ListContainerType.Dynamic,
      fixedListSize: 200
    } as IGeneralSettings,
    search: {
      isSearchOnKeyPressEnabled: false,
      debounceTime: 500
    } as ISearchSettings,
    pagination: {
      paginationType: PaginationType.Classic,
      maxVisiblePages: 3
    } as IPaginationSettings,
    theme: {
      backgroundColor: BackgroundColor.DarkGray,
      primaryColor: '#ff9900',
      bsTheme: BsThemes.Light,
      primaryColorTopCord: -35,
      primaryColorLefCord: 81
    } as IThemeSettings
  } as ISettings
} as IState;