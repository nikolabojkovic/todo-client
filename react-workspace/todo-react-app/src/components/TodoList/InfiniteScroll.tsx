import { ReactNode, useEffect, useRef, useState } from "react";
import { delay, Subject } from "rxjs";

import { useTodoList } from "../../context";
import { Loader, LoadMore } from "..";
import { ListContainerType } from "../../models";

type Props = {
  children: ReactNode,
  updateEndIndex: (index: number) => void,
  fetch: Subject<number>,
  itemHeight: number
}

export function InfiniteScroll({children, updateEndIndex, fetch, itemHeight}: Props) {
  const todoList = useTodoList();

  const scrollableContainerRef = useRef(null as any);

  const [endIndex, setEndIndex] = useState(calculateEndIndex());
  const [isLoading, setIsLoading] = useState(false);

  const hasMoreItemsToLoad = !isLoading && (endIndex < todoList.displayList.length);

  useEffect(() => {
    updateEndIndex(calculateEndIndex());
    const subscription = fetch
      .pipe(
        delay(1000),
      )
      .subscribe((endIndex: number) => {
        setEndIndex(endIndex);
        setIsLoading(false);
        updateEndIndex(endIndex);
      });

    return () => subscription.unsubscribe();
  }, []);

  useEffect(() => {
    const onDocumentScroll = () => {
      if (todoList.settings.general.listSizeType !== ListContainerType.Dynamic) return;

      const { scrollTop, clientHeight, scrollHeight } = document.documentElement;
      handleScroll (scrollTop, clientHeight, scrollHeight);
    };

    window.addEventListener("scroll", onDocumentScroll);
    return () => {
      window.removeEventListener("scroll", onDocumentScroll);
    };
  });

  useEffect(() => {
    const endIndex = calculateEndIndex();
    updateEndIndex(endIndex);
    setEndIndex(endIndex);
    setIsLoading(false);
  }, [todoList.settings.general.listSizeType]);

  const handleScroll = (scrollTop: number, clientHeight: number, scrollHeight: number) => {
    if (isLoading) {
      return;
    }

    if ((scrollTop + clientHeight >= scrollHeight - 20) 
     && !isLoading 
     && hasMoreItemsToLoad) {
      setIsLoading(true);
      fetch.next(endIndex + 5);
    }    
  };
  
  function calculateEndIndex(): number {
    const endIndex = (todoList.settings.general.listSizeType === ListContainerType.Fixed  
      ? ((todoList.settings.general.fixedListSize) / itemHeight)
      : (document.documentElement.clientHeight - 200) / itemHeight);

    return endIndex;
  }
  
  return (
    <div 
      id="scrollable-container"  
      data-testid="scrollable-container"
      ref={scrollableContainerRef}
      onScroll={() => {
        if (todoList.settings.general.listSizeType !== ListContainerType.Fixed) return;

        handleScroll(scrollableContainerRef.current.scrollTop, scrollableContainerRef.current.clientHeight, scrollableContainerRef.current.scrollHeight);
      }}     
      style={ 
        todoList.settings.general.listSizeType === ListContainerType.Fixed 
        ? { 
            height: todoList.settings.general.fixedListSize,
            overflow: 'overlay'
          } : {}
        }
      
    >
      { children }
      { isLoading &&
        <Loader height={150} />
      }  
      { hasMoreItemsToLoad && 
        <LoadMore />
      } 
    </div>
  );
}