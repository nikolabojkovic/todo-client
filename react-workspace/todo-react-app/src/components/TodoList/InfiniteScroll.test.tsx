import renderer, { act } from 'react-test-renderer';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import { TodosContext, TodosDispatchContext, stateTestData } from "../../context";
import { InfiniteScroll } from './InfiniteScroll';
import { IGeneralSettings, ISettings, ListContainerType } from '../../models';
import { of, Subject } from 'rxjs';

describe('InfiniteScroll', () => {
  const globalContext = {
    state: {
      ...stateTestData,
    },
    dispatch: jest.fn()
  };
  
  it('match snapshot infinite scroll fixed list size', () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false,
        settings: {
          ...globalContext.state.settings,
          general: {
            ...globalContext.state.settings.general,
            listSizeType: ListContainerType.Fixed
          } as IGeneralSettings
        } as ISettings
      }
    };
    const updateEndIndex = jest.fn();
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
          <InfiniteScroll 
            updateEndIndex={updateEndIndex} 
            fetch={new Subject<number>()} 
            itemHeight={55}
          >
            <div id="todo-list-content">
              <div>Item 1</div>
              <div>Item 2</div>
              <div>Item 3</div>
              <div>Item 4</div>
              <div>Item 5</div>
              <div>Item 6</div>
              <div>Item 7</div>
              <div>Item 8</div>
              <div>Item 9</div>
              <div>Item 10</div>
            </div>
          </InfiniteScroll>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    const tree = renderer.create(
      jsxElement
    ).toJSON();
    render(jsxElement);
    expect(tree).toMatchSnapshot();
  });

  it('match snapshot infinite scroll dynamic list size', () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false,
        settings: {
          ...globalContext.state.settings,
          general: {
            ...globalContext.state.settings.general,
            listSizeType: ListContainerType.Dynamic
          } as IGeneralSettings
        } as ISettings
      }
    };
    const updateEndIndex = jest.fn();
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <InfiniteScroll 
              updateEndIndex={updateEndIndex} 
              fetch={new Subject<number>()} 
              itemHeight={55}
            >
            <div id="todo-list-content">
              <div>Item 1</div>
              <div>Item 2</div>
              <div>Item 3</div>
              <div>Item 4</div>
              <div>Item 5</div>
              <div>Item 6</div>
              <div>Item 7</div>
              <div>Item 8</div>
              <div>Item 9</div>
              <div>Item 10</div>
            </div>
          </InfiniteScroll>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    const tree = renderer.create(
      jsxElement
    ).toJSON();
    render(jsxElement);
    expect(tree).toMatchSnapshot();
  });

  it('should scroll to bottom on fixed list', async () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false,
        settings: {
          ...globalContext.state.settings,
          general: {
            ...globalContext.state.settings.general,
            listSizeType: ListContainerType.Fixed
          } as IGeneralSettings
        } as ISettings
      }
    };
    const updateEndIndex = jest.fn();
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <section className="App__todo-list">
                <InfiniteScroll 
                  updateEndIndex={updateEndIndex} 
                  fetch={new Subject<number>()} 
                  itemHeight={55}
                >
                  <div data-testid="todo-list-content">
                    <div>Item 1</div>
                    <div>Item 2</div>
                    <div>Item 3</div>
                    <div>Item 4</div>
                    <div>Item 5</div>
                    <div>Item 6</div>
                    <div>Item 7</div>
                    <div>Item 8</div>
                    <div>Item 9</div>
                    <div>Item 10</div>
                  </div>   
                </InfiniteScroll>    
            </section>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    render(jsxElement);

    const todoList = screen.getByTestId("scrollable-container");
    fireEvent.scroll(todoList, { target: { scrollY: 1000 } });

    await waitFor(() => {
      expect(updateEndIndex).toBeCalled();
    }); 
  });

  it('scroll should trigger fetch items', async () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false,
        settings: {
          ...globalContext.state.settings,
          general: {
            ...globalContext.state.settings.general,
            listSizeType: ListContainerType.Fixed
          } as IGeneralSettings
        } as ISettings
      }
    };
    const updateEndIndex = jest.fn();
    const fetch = new Subject<number>();
    jest.spyOn(fetch, 'next').mockImplementation();
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <section className="App__todo-list">
                <InfiniteScroll 
                  updateEndIndex={updateEndIndex} 
                  fetch={fetch} 
                  itemHeight={55}
                >
                  <div data-testid="todo-list-content">
                    <div>Item 1</div>
                    <div>Item 2</div>
                    <div>Item 3</div>
                    <div>Item 4</div>
                    <div>Item 5</div>
                    <div>Item 6</div>
                    <div>Item 7</div>
                    <div>Item 8</div>
                    <div>Item 9</div>
                    <div>Item 10</div>
                  </div>   
                </InfiniteScroll>
            </section>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    await act( async () => {
      render(jsxElement);
    });

    const todoList = screen.getByTestId("scrollable-container");
    fireEvent.scroll(todoList, { target: { scrollY: 1000 } });

    await waitFor(() => {
      expect(fetch.next).toBeCalledWith(8.636363636363637);
      expect(fetch.next).toBeCalledTimes(1);
    }); 
  });

  it('scroll should not trigger fetch items', async () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false,
        settings: {
          ...globalContext.state.settings,
          general: {
            ...globalContext.state.settings.general,
            listSizeType: ListContainerType.Dynamic
          } as IGeneralSettings
        } as ISettings
      }
    };
    const updateEndIndex = jest.fn();
    const fetch = new Subject<number>();
    jest.spyOn(fetch, 'next').mockImplementation();
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <section className="App__todo-list">
                <InfiniteScroll 
                  updateEndIndex={updateEndIndex} 
                  fetch={fetch} 
                  itemHeight={55}
                >
                  <div data-testid="todo-list-content">
                    <div>Item 1</div>
                    <div>Item 2</div>
                    <div>Item 3</div>
                    <div>Item 4</div>
                    <div>Item 5</div>
                    <div>Item 6</div>
                    <div>Item 7</div>
                    <div>Item 8</div>
                    <div>Item 9</div>
                    <div>Item 10</div>
                  </div>   
                </InfiniteScroll>
            </section>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    await act( async () => {
      render(jsxElement);
    });

    const todoList = screen.getByTestId("scrollable-container");
    fireEvent.scroll(todoList, { target: { scrollY: 1000 } });

    expect(fetch.next).not.toBeCalledWith(8.636363636363637);
  });

  it('shold not invoke infinite scroll if loading is in progress', async () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false,
        settings: {
          ...globalContext.state.settings,
          general: {
            ...globalContext.state.settings.general,
            listSizeType: ListContainerType.Fixed,
            // fixedListSize: 1000
          } as IGeneralSettings
        } as ISettings
      }
    };
    const updateEndIndex = jest.fn();
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <section className="App__todo-list">
                <InfiniteScroll 
                  updateEndIndex={updateEndIndex} 
                  fetch={new Subject<number>()} 
                  itemHeight={55}
                >
                  <div data-testid="todo-list-content">
                    <div>Item 1</div>
                    <div>Item 2</div>
                    <div>Item 3</div>
                    <div>Item 4</div>
                    <div>Item 5</div>
                    <div>Item 6</div>
                    <div>Item 7</div>
                    <div>Item 8</div>
                    <div>Item 9</div>
                    <div>Item 10</div>
                  </div>   
                </InfiniteScroll>    
            </section>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    render(jsxElement);

    const todoList = screen.getByTestId("scrollable-container");
    fireEvent.scroll(todoList, { target: { scrollY: 1000 } });

    await waitFor(() => {
      expect(updateEndIndex).toBeCalled();
    }); 
    fireEvent.scroll(todoList, { target: { scrollY: 1000 } });
  });
});