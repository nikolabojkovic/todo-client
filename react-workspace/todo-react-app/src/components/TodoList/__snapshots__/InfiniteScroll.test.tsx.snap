// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll match snapshot infinite scroll dynamic list size 1`] = `
<div
  data-testid="scrollable-container"
  id="scrollable-container"
  onScroll={[Function]}
  style={Object {}}
>
  <div
    id="todo-list-content"
  >
    <div>
      Item 1
    </div>
    <div>
      Item 2
    </div>
    <div>
      Item 3
    </div>
    <div>
      Item 4
    </div>
    <div>
      Item 5
    </div>
    <div>
      Item 6
    </div>
    <div>
      Item 7
    </div>
    <div>
      Item 8
    </div>
    <div>
      Item 9
    </div>
    <div>
      Item 10
    </div>
  </div>
  <div
    className="App__todo-list__infinite-scroll-end"
    id="infinite-scroll-end"
  >
    <div
      className="App__todo-list__infinite-scroll-end--bouncing"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-down "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M169.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 274.7 54.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <div>
        Scroll down to load more
      </div>
    </div>
  </div>
</div>
`;

exports[`InfiniteScroll match snapshot infinite scroll fixed list size 1`] = `
<div
  data-testid="scrollable-container"
  id="scrollable-container"
  onScroll={[Function]}
  style={
    Object {
      "height": 200,
      "overflow": "overlay",
    }
  }
>
  <div
    id="todo-list-content"
  >
    <div>
      Item 1
    </div>
    <div>
      Item 2
    </div>
    <div>
      Item 3
    </div>
    <div>
      Item 4
    </div>
    <div>
      Item 5
    </div>
    <div>
      Item 6
    </div>
    <div>
      Item 7
    </div>
    <div>
      Item 8
    </div>
    <div>
      Item 9
    </div>
    <div>
      Item 10
    </div>
  </div>
  <div
    className="App__todo-list__infinite-scroll-end"
    id="infinite-scroll-end"
  >
    <div
      className="App__todo-list__infinite-scroll-end--bouncing"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-down "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M169.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 274.7 54.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <div>
        Scroll down to load more
      </div>
    </div>
  </div>
</div>
`;
